state/presence/presence.go

```diff
@@ -478,40 +491,131 @@ func (w *Watcher) sync() error {
+// Keys that we find are now alive will have a 'found alive' event queued.
+func (w* Watcher) lookupUnknownSeqs(unknownSeqs []int64, dead map[int64]bool, session *mgo.Session) error {
+	if len(unknownSeqs) == 0 {
+		// Nothing to do, with nothing unknown.
+		return nil
+	}
+	// We do cache *all* beingInfos, but they're reasonably small
+	seqToBeing := make(map[int64]beingInfo, len(unknownSeqs))
+	startTime := time.Now()
+	beingsC := w.beings.With(session)
+	remaining := unknownSeqs
+	for len(remaining) > 0 {
+		// batch this into reasonable lengths
+		// testing shows that it works just fine at 50,000 ids, but be a
+		// bit more conservative
+		batch := remaining
+		if len(remaining) > lookupBatchSize {
+			batch = remaining[:lookupBatchSize]
+			remaining = remaining[lookupBatchSize:]
+		} else {
+			remaining = nil
+		}
+		docIds := make([]string, len(batch))
+		for _, seq := range batch {
+			docIds = append(docIds, docIDInt64(w.modelUUID, seq))
+		}
+		query := beingsC.Find(bson.M{"_id": bson.M{"$in": docIds}})
+		// We don't need the _id returned, as its just a way to lookup the seq,
+		// and _id is quite large
+		query = query.Select(bson.M{"_id": false, "key": true, "seq": true})
+		query.Batch(lookupBatchSize)
+		beingIter := query.Iter()
+		being := beingInfo{}
+		for beingIter.Next(&being) {
+			seqToBeing[being.Seq] = being
+		}
+		if err := beingIter.Close(); err != nil {
+			if err != mgo.ErrNotFound {
+				// This may be an old sequence, not considered fatal
+				return err
+			}
+		}
+	}
+	rate := ""
+	elapsed := time.Since(startTime)
+	if len(unknownSeqs) > 0 {
+		seqPerMS :=  float64(len(unknownSeqs)) / (elapsed.Seconds() * 1000.0)
+		rate = fmt.Sprintf(" (%.1fseq/ms)", seqPerMS)
+	}
+	unownedCount := 0
+	for _, seq := range unknownSeqs {
+		being, ok := seqToBeing[seq]
+		if !ok {
+			// Not Found
+			unownedCount++
+			logger.Tracef("[%s] found seq=%d unowned", w.modelUUID[:6], seq)
+			continue
+		}
+		cur := w.beingSeq[being.Key]
+		if cur < seq {
+			delete(w.beingKey, cur)
+		} else {
+			// We already have a sequence for this key, and it is
+			// newer than the one we just saw.
+			continue
+		}
+		// Start tracking the new sequence for this key
+		w.beingKey[seq] = being.Key
+		w.beingSeq[being.Key] = seq
+		if cur > 0 || dead[seq] {
+			// if cur > 0, then we already think this is alive, no
+			// need to queue another message.
+			// if dead[] then we still wouldn't queue an alive message
+			// because we are writing a 'is dead' message.
+			continue
+		}
+		logger.Tracef("[%s] found seq=%d alive with key %q", w.modelUUID[:6], seq, being.Key)
+		for _, ch := range w.watches[being.Key] {
+			w.pending = append(w.pending, event{ch, being.Key, true})
+		}
+	}
+	// TODO(jam): 2017-04-18 This runs every 30s, probably needs to be Trace...
+	logger.Debugf("looked up %d unknown sequences for %q (%d unowned) in %v%s from %q",
+		len(unknownSeqs), w.modelUUID, unownedCount, elapsed, rate, beingsC.Name)
```

Comment:
Yeah, having run it on an active controller, we see ~5 unknown sequences every cycle, so it does print every 30s. I was thinking that it should have reached stasis, in which case we shouldn't be doing this lookup each time.